EASY-2244 add deposit.origin property to easy-deposit-properties

jo-pol opened the pull request:
Fixes EASY-2244 add deposit.origin property to easy-deposit-properties

#### When applied it will
* 
* 
* 

#### By submitting this pull request I confirm that
* [x] all files contain licenses (`mvn license:format`)
* [x] the project compiles (`mvn clean install -DskipTests=true`)
* [x] all unit tests pass (`mvn test`)
* [x] the changes have been tested on `deasy`

#### Where should the reviewer @DANS-KNAW/easy start?

#### How should this be manually tested?

Submitted a new deposit with the deposit-ui, subsequently the next commands
```
easy-deposit-properties load-props --doUpdate /var/opt/dans.knaw.nl/tmp/easy-deposit-api/drafts/user001/d062895f-a289-411b-bef4-9e4ca9da23e3/deposit.properties
easy-deposit-properties load-props --doUpdate /var/opt/dans.knaw.nl/tmp/easy-ingest-flow-inbox/5afce5ef-7f91-466a-8552-1f61d521a0cd/deposit.properties
```
Results: `is not writable`, ` Cannot insert this identifier: identifier urn already exists for depositId.`

#### Related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

jo-pol committed:
followed issue instructions up to and including 7b

jo-pol committed:
getting it compiled

jo-pol committed:
fixed some unit test breaking on an irrelevant order

jo-pol committed:
Merge remote-tracking branch 'blessed/master' into 2244-origin-column

rvanheest commented on docs/schema.graphql:
```diff
@@ -1,1530 +1,1053 @@
-input AddBagNameInput {
-  """The deposit's identifier."""
-  depositId: UUID!
+# This file was generated based on ".graphqlconfig". Do not edit manually.
```
Please refer to https://github.com/DANS-KNAW/easy-deposit-properties#graphql-schema for updating the schema. Don't use IntelliJ's generator!

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/graphql/types/MutationType.scala:
```diff
@@ -66,6 +67,14 @@ trait MutationType {
     astDirectives = Vector.empty,
     astNodes = Vector.empty,
   )
+  private val originOptionInputField: InputField[Option[Origin.Value]] = InputField(
```
Not sure if these types line up. This has `Option[Origin]`, while below you do `.asInstanceOf[Origin]`.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/CommonResultSetParsers.scala:
```diff
@@ -62,7 +62,8 @@ private[sql] trait CommonResultSetParsers {
       creationTimestamp <- parseDateTime(resultSet.getTimestamp("creationTimestamp", timeZone), timeZone)
       depositorId = resultSet.getString("depositorId")
-    } yield Deposit(depositId, bagName, creationTimestamp, depositorId)
+      origin = Origin.withName(resultSet.getString("origin"))
```
This probably won't work on `deasy` with PostgreSQL. See EASY-2300. If you can fix it here, perhaps you can fix the bug in that issue as well?

rvanheest commented on src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/SQLDepositDaoSpec.scala:
```diff
@@ -237,7 +237,8 @@ class SQLDepositDaoSpec extends TestSupportFixture
     val depositId6 = UUID.fromString("00000000-0000-0000-0000-000000000006")
-    val deposit6 = Deposit(depositId6, Some(""), new DateTime(2019, 6, 6, 0, 0, timeZone), "user003")
+    val deposit6 = Deposit(depositId6, Some(""), new DateTime(2019, 6, 6, 0, 0, timeZone), "user003", Origin.API
+    )
```
keep it on the same line, please 😜

rvanheest commented on src/test/scala/nl.knaw.dans.easy.properties/server/GraphQLResolveSpec.scala:
```diff
@@ -1527,6 +1531,8 @@ class GraphQLResolveSpec extends TestSupportFixture
     runQuery(input)
   }
 
+
```
no need for these empty lines

rvanheest commented on src/test/scala/nl.knaw.dans.easy.properties/server/GraphQLResolveSpec.scala:
```diff
@@ -1553,6 +1559,11 @@ class GraphQLResolveSpec extends TestSupportFixture
   }
 
+  private def inAnyOrder[T](expected: Seq[T]): FunctionAdapter1[Seq[T], Boolean] = {
+    // TODO vararch? To some package or find in library.
```
or find out why it doesn't require the `inAnyOrder` on my machine nor on Travis, but why it does require it on your end 😉

jo-pol replied:
The tests run with blessed/master. So it is caused by my changes, or problems I introduced.

rvanheest replied:
If it's caused by your change, let's for now just revert this `inAnyOrder`, put the elements in the order that the test expects them to be in and put a TODO somewhere to indicate that we need to take a good look at it. To be completely honest with you, I remember having to deal with this issue before while writing these tests and as you may have seen, some of them already use this proposed 'quick fix'. I've had the same problem and this is how I solved it. I'm not particularly worried about this ordering thing, especially given that we're talking about mocks here.

rvanheest commented:
I think I miss a diff in `src/test/resources/database/{db.script, db.properties}`. See the README in that directory for all details!

jo-pol committed:
all tests now get executed, 57 out of 487 are broken

jo-pol committed:
executed generateHsqldbScripts.sh

jo-pol committed:
executed generateHsqldbScripts.sh and more review comments

jo-pol committed:
regenerated docs/schema.graphql as described in readme

jo-pol committed:
10 broken tests down, 47 to go (about 2 types)

jo-pol committed:
22 tests to go

rvanheest commented on src/main/assembly/dist/install/db-tables.sql:
```diff
@@ -1,8 +1,11 @@
+CREATE TYPE Origin AS ENUM ('SWORD2', 'API', 'SMD');
```
This is not gonna work; see EASY-2300.

My current proposal for that bug (which is similar to the one you introduced here) can be found at https://github.com/DANS-KNAW/easy-deposit-properties/pull/5

jo-pol committed:
3 more tests down

jo-pol committed:
1 more test down

jo-pol committed:
fixed the last unit tests

jo-pol committed:
Merge remote-tracking branch 'blessed/master' into 2244-origin-column

# Conflicts:
#	src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala
#	src/test/resources/database/db.properties
#	src/test/resources/database/db.script

jo-pol committed:
as #5 for origin

jo-pol committed:
generated src/test/resources/database/db.*

in one terminal
  mvn clean install -DskipTests=true && run-service.sh
then in another
  get-graphql-schema http://localhost:20200/graphql > docs/schema.graphql && ./src/test/resources/database/generateHsqldbScripts.sh

jo-pol committed:
fixed unit test

rvanheest commented on README.md:
```diff
@@ -70,7 +70,7 @@ To generate the latest version of the GraphQL schema for `easy-deposit-propertie
     npm install -g get-graphql-schema
     
-    # start the service
+    # (re)start the service (after: mvn clean install -DskipTests=true): see GRAPHQL INTERFACE above
```
I think you can even make this a link...

```suggestion
    # (re)start the service (after: mvn clean install -DskipTests=true): see [GRAPHQL INTERFACE](#graphql-interface)
```

**Please test this yourself!!!**

jo-pol replied:
You are mixing up code blocks and plain markdown. The latter is not applied withing the first.

rvanheest commented on docs/schema.graphql:
```diff
@@ -1067,6 +1087,18 @@ enum OrderDirection {
 }
 
+"""The origin of the deposit."""
+enum Origin {
```
The documentation isn't right for these elements... Should be changed in the Sangria code, after which you should generate a new version of this file!

rvanheest commented on src/main/assembly/dist/install/db-tables.sql:
```diff
@@ -2,7 +2,8 @@ CREATE TABLE Deposit (
     creationTimestamp TIMESTAMP WITH TIME ZONE NOT NULL,
-    depositorId VARCHAR(64) NOT NULL
+    depositorId VARCHAR(64) NOT NULL,
+    origin VARCHAR(64) NOT NULL
```
Now that I think of it, isn't `VARCHAR(64)` a bit much? The values (for now) are only `SWORD2`, `SMD` or `API`. Perhaps a smaller space should suffice? LMKWYT

jo-pol replied:
... and inconsistent with test/resources/database/database.sql

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/repository/package.scala:
```diff
@@ -31,7 +32,11 @@ package object repository {
-    def apply(ts: Seq[Throwable]): InvalidValueError = InvalidValueError(ts.map(_.getMessage).mkString("\n"))
+    def apply(ts: Seq[Throwable], query: String =""): InvalidValueError = {
+      val msg = ts.map(_.getMessage).mkString("\n")
+      logger.debug(s"$msg [$query]")
```
if `query` is empty, you get `"<some msg> []"` here.

jo-pol replied:
It was a quick hack to figure out why a test failed. Should either drop it or redesign exception classes in general to provide a little more context to simplify trouble shooting in the future. For now renamed argument to debugContext.

rvanheest replied:
Probably drop it for now and if you'd like, try to redesign the exceptions classes in another PR (rather than this one).

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/repository/package.scala:
```diff
@@ -31,7 +32,11 @@ package object repository {
   case class InvalidValueError(override val msg: String) extends QueryError(msg)
   object InvalidValueError {
-    def apply(ts: Seq[Throwable]): InvalidValueError = InvalidValueError(ts.map(_.getMessage).mkString("\n"))
+    def apply(ts: Seq[Throwable], query: String =""): InvalidValueError = {
```
```suggestion
    def apply(ts: Seq[Throwable], query: String = ""): InvalidValueError = {
```

rvanheest commented on src/test/resources/database/database.sql:
```diff
@@ -6,7 +6,8 @@ CREATE TABLE Deposit (
     creationTimestamp TIMESTAMP WITH TIME ZONE NOT NULL,
-    depositorId VARCHAR(64) NOT NULL
+    depositorId VARCHAR(64) NOT NULL,
+    origin VARCHAR(6) NOT NULL,
```
Is this trailing comma allowed in SQL? I'm fairly sure it isn't in PostgreSQL.

```suggestion
    origin VARCHAR(6) NOT NULL
```

jo-pol replied:
If not allowed, then why do the tests succeed?

rvanheest commented on src/test/resources/graphql-examples/deposit/addDeposit/addExistingDeposit.graphql:
```diff
@@ -1,5 +1,5 @@
-    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000004", bagName: "bag4", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001"}) {
+    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000004", bagName: "bag4", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001", origin: API}) {
         clientMutationId
         deposit {
```
perhaps you could add the `origin` field to the output as well (also causes the `*.json` file to change)

Same for the other examples in `addDeposit` below, please!

rvanheest reviewed (changes requested):
Here's some more feedback.
Awesome to see you got all the tests working 😉

jo-pol committed:
review comments

jo-pol committed:
enum comments

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/legacyImport/ImportProps.scala:
```diff
@@ -171,6 +163,20 @@ class ImportProps(repository: Repository, interactor: Interactor, datacite: Data
     InputState(label, description, timestamp)
   }
 
+  private def getOrAskEnumProp(enum: Enumeration, propertyKey: FeedBackMessage, propertyDescription: FeedBackMessage, props: PropertiesConfiguration, depositId: DepositId): enum.Value = {
```
are there only two usages of this extraction, or could more of them be extracted? Looks like a very and reusable piece of code!

jo-pol replied:
In this source yes. Another pull request for other instances.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.properties/app/legacyImport/ImportProps.scala:
```diff
@@ -25,14 +25,15 @@ import cats.syntax.option._
 import cats.syntax.traverse._
 import nl.knaw.dans.easy.properties.ApplicationErrorOr
 import nl.knaw.dans.easy.properties.Command.FeedBackMessage
+import nl.knaw.dans.easy.properties.app.model.Origin.Origin
```
unused import

rvanheest commented on src/test/resources/graphql-examples/deposit/addDeposit/addNewDeposit.graphql:
```diff
@@ -1,5 +1,5 @@
 mutation {
-    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000006", bagName: "bag6", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001"}) {
+    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000006", bagName: "bag6", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001", origin: SWORD2}) {
```
please add the `origin` in the response query as well (this will also change the expected JSON result)

rvanheest commented on src/test/resources/graphql-examples/deposit/addDeposit/addNewDepositNoBagName.graphql:
```diff
@@ -1,5 +1,5 @@
 mutation {
-    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000006", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001"}) {
+    addDeposit(input: {clientMutationId: "my-mutation-id", depositId: "00000000-0000-0000-0000-000000000006", creationTimestamp: "2019-05-15T23:36:00.000Z", depositorId: "archie001", origin: SMD}) {
```
please add the `origin` in the response query as well (this will also change the expected JSON result)

rvanheest replied:
Please also include the `origin` field in the following example queries:

* `deposit/findDeposit/plain.graphql`
* `deposits/listAllDeposits/*.graphql`

jo-pol committed:
fix legacy import, review comments

jo-pol committed:
replaced inAnyOrder method by changing the expected order

jo-pol committed:
added origin to graphql-examples/deposits/listAllDeposits

jo-pol committed:
Merge remote-tracking branch 'blessed/master' into 2244-origin-column

# Conflicts:
#	src/main/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGenerator.scala
#	src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala

jo-pol committed:
typo when fixing merge conflicts
